plugins/pixart-rf/fu-pxi-wireless-device.c

```diff
@@ -0,0 +1,606 @@
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static gboolean
+fu_pxi_wireless_device_get_feature (FuDevice *self, guint8 *buf, guint bufsz, GError **error)
+{
+#ifdef HAVE_HIDRAW_H
+	if (!fu_udev_device_ioctl (FU_UDEV_DEVICE (self),
+				   HIDIOCGFEATURE(bufsz), buf,
+				   NULL, error)) {
+		return FALSE;
+	}
+	if (g_getenv ("FWUPD_PIXART_RF_VERBOSE") != NULL)
+		fu_common_dump_raw (G_LOG_DOMAIN, "GetFeature", buf, bufsz);
+	return TRUE;
+#else
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_NOT_SUPPORTED,
+			     "<linux/hidraw.h> not available");
+	return FALSE;
+#endif
+}
+
+static FuPxiReceiverDevice *
+fu_pxi_wireless_device_get_parent (FuDevice *self, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent (FU_DEVICE (self));
+	if (parent == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "no parent set");
+		return NULL;
+	}
+	return FU_PXI_RECEIVER_DEVICE (FU_UDEV_DEVICE (parent));
+}
+
+static gboolean
+fu_pxi_wireless_device_open (FuDevice *device, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent (device);
+	if (parent == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "no parent device");
+		return FALSE;
+	}
+	return fu_device_open (parent, error);
+}
+
+static gboolean
+fu_pxi_wireless_device_close (FuDevice *device, GError **error)
+{
+	FuDevice *parent = fu_device_get_parent (device);
+	if (parent == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "no parent device");
+		return FALSE;
+	}
+	return fu_device_close (parent, error);
+}
+
+static gboolean
+fu_pxi_wireless_device_get_cmd_response (FuPxiWirelessDevice *device,
+					 guint8 *buf, guint bufsz,
+					 GError **error)
+{
+	FuPxiReceiverDevice *parent = fu_pxi_wireless_device_get_parent (FU_DEVICE (device) ,error);
+	guint16 retry = 0;
+	guint8 status = 0x0;
+
+	while (1) {
+		guint8 sn = 0x0;
+		memset (buf, 0, bufsz);
+		buf[0] = PXI_HID_WIRELESS_DEV_OTA_REPORT_ID;
+
+		g_usleep (5 * 1000);
+
+		if (!fu_pxi_wireless_device_get_feature (FU_DEVICE (parent), buf, bufsz, error))
+			return FALSE;
+
+		if (!fu_common_read_uint8_safe (buf, bufsz, 0x4, &sn, error))
+			return FALSE;
+
+		if (device->sn != sn)
+			retry++;
+		else
+			break;
+
+		if (retry == FU_PXI_WIRELESS_DEVICE_RETRY_MAXIMUM) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_READ,
+				     "reach retry maximum hid sn fail, got 0x%04x, expected 0x%04x",
+				     sn, device->sn);
+			return FALSE;
+		}
+
+		/*if wireless device not reply to receiver, keep to wait */
+		if (!fu_common_read_uint8_safe (buf, bufsz, 0x5, &status, error))
+			return FALSE;
+		if (status == OTA_RSP_NOT_READY) {
+			retry = 0x0;
+			g_debug ("OTA_RSP_NOT_READY");
+		}
```

Comment:
I don't really follow how this code block ever gets hit.  Shouldn't getting a feature and reading uint8 both have failed if it didn't reply?

Reply:
the wireless could be busy for other command ,so we need to poll the device.

Reply:
but so if it's busy with another command, you can still read the status byte?  If so, no concerns.